### PR TITLE
[rhi] MetalSurface functions

### PR DESCRIPTION
**Pull request by AntonioFerreras**
Issue: #

### <samp>🤖 Generated by Copilot at a89e235</samp>

This pull request adds the ability to create and use a Metal surface for rendering graphics using the RHI interface. It introduces the `MetalSurface` class in `metal_device.h` and `metal_device.mm`, and links the QuartzCore framework in `CMakeLists.txt`.

### <samp>🤖 Generated by Copilot at a89e235</samp>

*  Add the MetalSurface class and the create_surface method to support rendering on Metal ([link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-5b304e188996abd217fad85fd8b2434e729ad9c089eb9ef48a848c0fcc74d614R249-R292), [link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-5b304e188996abd217fad85fd8b2434e729ad9c089eb9ef48a848c0fcc74d614L260-R310), [link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-c350a27659df1ebcf2c854e2c21a80f82de87d3b325a80e00f6a6dd6ca53303bR608-R682), [link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-c350a27659df1ebcf2c854e2c21a80f82de87d3b325a80e00f6a6dd6ca53303bR707-R711))
*  Add the QuartzCore framework and header to use the CAMetalLayer and CAMetalDrawable classes for the MetalSurface ([link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-464e3a16b2cf83d04b813312367cf6b4d56b4d313af162dde3237c8352374667L21-R21), [link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-5b304e188996abd217fad85fd8b2434e729ad9c089eb9ef48a848c0fcc74d614L13-R18), [link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-5b304e188996abd217fad85fd8b2434e729ad9c089eb9ef48a848c0fcc74d614L29-R36))
*  Add helper functions to convert RHI enums to Metal enums for buffer format, image dimension, and image usage ([link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-c350a27659df1ebcf2c854e2c21a80f82de87d3b325a80e00f6a6dd6ca53303bR10-R84))
*  Add a TODO comment to the MetalShaderResourceSet class to indicate the need for raster resources support ([link](https://github.com/taichi-dev/taichi/pull/8274/files?diff=unified&w=0#diff-5b304e188996abd217fad85fd8b2434e729ad9c089eb9ef48a848c0fcc74d614L174-R180))

### CODE REVIEW

File: taichi/rhi/metal/metal_device.h [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
+#include "taichi/rhi/metal/metal_api.h"
+#include <memory>
 
 #if defined(__APPLE__) && defined(__OBJC__)
+#import <CoreGraphics/CoreGraphics.h>
 #import <Foundation/Foundation.h>
 #import <Metal/Metal.h>
 #import <MetalKit/MetalKit.h>
-#import <CoreGraphics/CoreGraphics.h>
+#import <QuartzCore/QuartzCore.h>
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.h:13]: Why is this replaced?

Comment (bobcao3) [reply taichi/rhi/metal/metal_device.h:13]: nvm the order changed it appears

Comment (bobcao3) [reply taichi/rhi/metal/metal_device.h:13]: it's better to not change order of things unless it's a formatting / style related PR

File: taichi/rhi/metal/metal_device.h [hunk 2]
```diff
@@ -41,7 +47,7 @@ class MetalStream;
 class MetalDevice;
 
 struct MetalMemory {
- public:
+public:
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.h:50]: They should be formatted as before, one space before public and private

File: taichi/rhi/metal/metal_device.h (hunk 1 above)
Comment (AntonioFerreras) [reply taichi/rhi/metal/metal_device.h:13]: I don't know, but the CoreGraphics header is still included in that file, just earlier

File: taichi/rhi/metal/metal_device.h [hunk 3]
```diff
@@ -91,19 +97,17 @@ struct MetalWorkgroupSize {
   uint32_t z{0};
 };
 class MetalPipeline final : public Pipeline {
- public:
+public:
   // `mtl_library`, `mtl_function`, `mtl_compute_pipeline_state` should be
   // already retained.
-  explicit MetalPipeline(const MetalDevice &device,
-                         MTLLibrary_id mtl_library,
+  explicit MetalPipeline(const MetalDevice &device, MTLLibrary_id mtl_library,
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.h:103]: ditto, should be formatted as before

File: taichi/rhi/metal/metal_device.h [hunk 4]
```diff
@@ -145,37 +149,34 @@ struct MetalShaderResource {
...
   ShaderResourceSet &buffer(uint32_t binding, DeviceAllocation alloc) final;
 
-  ShaderResourceSet &image(uint32_t binding,
-                           DeviceAllocation alloc,
+  ShaderResourceSet &image(uint32_t binding, DeviceAllocation alloc,
                            ImageSamplerConfig sampler_config) override;
 
-  ShaderResourceSet &rw_image(uint32_t binding,
-                              DeviceAllocation alloc,
+  ShaderResourceSet &rw_image(uint32_t binding, DeviceAllocation alloc,
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.h:166]: ditto

File: taichi/rhi/metal/metal_device.h [hunk 5]
```diff
@@ -192,13 +193,12 @@ class MetalCommandList final : public CommandList {
   void buffer_fill(DevicePtr ptr, size_t size, uint32_t data) noexcept final;
   RhiResult dispatch(uint32_t x, uint32_t y = 1, uint32_t z = 1) noexcept final;
 
-  void image_transition(DeviceAllocation img,
-                        ImageLayout old_layout,
+  void image_transition(DeviceAllocation img, ImageLayout old_layout,
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.h:196]: ditto

File: taichi/rhi/metal/metal_device.h [hunk 6]
```diff
@@ -219,47 +219,78 @@ class MetalStream final : public Stream {
...
-      CommandList *cmdlist,
-      const std::vector<StreamSemaphore> &wait_semaphores = {}) final;
-  StreamSemaphore submit_synced(
-      CommandList *cmdlist,
-      const std::vector<StreamSemaphore> &wait_semaphores = {}) final;
+  StreamSemaphore
+  submit(CommandList *cmdlist,
+         const std::vector<StreamSemaphore> &wait_semaphores = {}) final;
+  StreamSemaphore
+  submit_synced(CommandList *cmdlist,
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.h:229]: ditto, formatting

File: taichi/rhi/metal/metal_device.mm [hunk 7]
```diff
@@ -530,6 +605,82 @@ DeviceCapabilityConfig collect_metal_device_caps(MTLDevice_id mtl_device) {
...
+
+void MetalSurface::present_image(
+    const std::vector<StreamSemaphore> &wait_semaphores) {
+
+  Stream *stream = device_->get_compute_stream();
+  auto [cmd_list, res] = stream->new_command_list_unique();
+  MetalCommandList *cmd_list2 = (MetalCommandList *)cmd_list.get();
+  RHI_ASSERT(res == RhiResult::success);
+
+  [cmd_list2->finalize() presentDrawable:current_drawable_];
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.mm:665]: It's probably cleaner to use the Metal API directly here instead of using the RHI version.

Comment (bobcao3) [reply taichi/rhi/metal/metal_device.mm:665]: Also after checking the API for a bit I think you don't need an command buffer for this. Here's an API:
https://developer.apple.com/documentation/metal/mtldrawable/1470284-present?language=objc

From the commandbuffer API it says the presentDrawable is just a convenience function wrapping this underlying `present` function

**Review by bobcao3 — APPROVED** [review]
LGTM

**Pull request merged**